### PR TITLE
[fields] Fix new index calculation on letter section

### PR DESCRIPTION
**Pull request by arthurbalduini**
- Fixes https://github.com/mui/mui-x/issues/13213
- Implements tests to assure no regressions and basic tests for the PageUp and PageDown support

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts
Comment (arthurbalduini) [inline packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts:326]: My first thought was to use Math.abs(), which I believe could also work as a fix, but after some thinking I concluded that this approach would be safer to assure that we will always return an index within the correct range.

Comment (flaviendelangle) [reply packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts:326]: I do agree that using the modulo seems safer than an absolute value if in the future for some reason we end up with a weird delta.

Nitpick on the naming, since the operation we are doing in a clamping (https://en.wikipedia.org/wiki/Clamping_(graphics))

```suggestion
    const clampedIndex = (newOptionIndex + options.length) % options.length;
```

**Review by flaviendelangle — APPROVED** [review]
I'll let @LukasTy have the final approval here, but for me it's looking great :partying_face:

File: packages/x-date-pickers/src/TimeField/tests/editing.TimeField.test.tsx
Comment (LukasTy) [inline packages/x-date-pickers/src/TimeField/tests/editing.TimeField.test.tsx:229]: Nitpick: Given [the definition](https://www.merriam-webster.com/dictionary/trespass) of `trespass`, WDYT about going with `flip` for simplicity? 🤔 
This would apply to all instances.

File: packages/x-date-pickers/src/TimeField/tests/editing.TimeField.test.tsx
Comment (LukasTy) [inline packages/x-date-pickers/src/TimeField/tests/editing.TimeField.test.tsx:331]: Missing test description. 🙈

File: packages/x-date-pickers/src/DateField/tests/editing.DateField.test.tsx
Comment (LukasTy) [reply packages/x-date-pickers/src/DateField/tests/editing.DateField.test.tsx:519]: Nitpick: adding tests for the day and year sections could also make sense. 🤔

**Review by LukasTy — APPROVED** [review]
Thank you for finding and taking care of this issue! 🙏 💯 🎉 
Great work! 👏

File: packages/x-date-pickers/src/DateField/tests/editing.DateField.test.tsx
Comment (LukasTy) [inline packages/x-date-pickers/src/DateField/tests/editing.DateField.test.tsx:741]: @flaviendelangle WDYT about making the `minimum` and `maximum` aware of the `minDate` and `maxDate`?
Didn't we do it only, because `shouldDisableYear` exists? 🤔 
IMHO, the current behavior seems a bit strange, especially given the differences in adapters. 🙈

**Review by LukasTy — APPROVED** [review]

**Pull request merged**
